test/pytest/test_bridge.py

```diff
@@ -655,6 +677,84 @@ async def test_fsreplace1_error(transport: MockTransport, tmp_path: Path) -> Non
+                                   'message': '"user" attribute is empty while "group" is provided'
+    })
+
+    await transport.check_open('fsreplace1', path=str(tmp_path / 'test'), attrs={'user': 'cockpit', 'group': []},
+                               problem='protocol-error',
+                               reply_keys={
+                                   'message': "attribute 'attrs': attribute 'group': must be a string or integer"
+    })
+
+    await transport.check_open('fsreplace1', path=str(tmp_path / 'test'), attrs={'user': [], 'group': 'test'},
+                               problem='protocol-error',
+                               reply_keys={
+                                   'message': "attribute 'attrs': attribute 'user': must be a string or integer"
+    })
+
+    mock_val = GrpPwMock(uid=0, gid=1000)
+    with unittest.mock.patch('pwd.getpwnam', side_effect=KeyError()):
+        await transport.check_open('fsreplace1', path=str(tmp_path / 'test'),
+                                   attrs={'user': 'bazinga', 'group': 'foo'},
+                                   problem='internal-error',
+                                   reply_keys={
+                                       'message': 'uid not found for bazinga'
+        })
+
+    with unittest.mock.patch('pwd.getpwnam', return_value=mock_val):
+        with unittest.mock.patch('grp.getgrnam', side_effect=KeyError()):
+            await transport.check_open('fsreplace1', path=str(tmp_path / 'test'),
+                                       attrs={'user': 'bazinga', 'group': 'test'},
+                                       problem='internal-error',
+                                       reply_keys={
+                                           'message': 'gid not found for test'
+            })
+
+
+ATTRS_TEST_DATA = [
+    (GrpPwMock(1111, 1110), {'user': 1111, 'group': 1110}),
+    (GrpPwMock(1111, 1110), {'user': 'monkey', 'group': 'group'}),
+]
+
+
+@pytest.mark.parametrize(('user_group_mock_arg', 'attrs'), ATTRS_TEST_DATA)
+@pytest.mark.asyncio
+async def test_fsreplace1_attrs(transport: MockTransport, fchown_mock: unittest.mock.MagicMock,
+                                tmp_path: Path, user_group_mock, attrs) -> None:
+    async def create_file_with_attrs(filename: str, attrs: Dict[str, Union[int, str]]) -> None:
+        test_file = str(tmp_path / filename)
+        ch = await transport.check_open('fsreplace1', path=str(test_file), attrs=attrs)
+        transport.send_data(ch, b'content')
+        transport.send_done(ch)
+        await transport.assert_msg('', command='done', channel=ch)
+        await transport.check_close(ch)
+
+    await create_file_with_attrs('test', attrs)
+    fchown_mock.assert_called()
+    _, call_uid, call_gid = fchown_mock.call_args[0]
+    assert call_uid == user_group_mock.pw_uid
+    if user_group_mock.gr_gid is None:
+        assert call_gid == user_group_mock.pw_uid
+    else:
+        assert call_gid == user_group_mock.gr_gid
```

Comment:
Please also check replacing an existing file with new user/group, to make sure the attrs one wins over "copy permissions from existing files". This requires some more care in the mock check, i.e. you need to ensure how often it was called.